Pidgin.Examples/Expr/ExprParser.cs

```diff
@@ -0,0 +1,182 @@
+        public static readonly Parser<char, string> Else = String("else").Labelled("else");
+        public static readonly Parser<char, string> Then = String("then").Labelled("then");
+        public static readonly Parser<char, string> In = String("in").Labelled("in");
+
+        private static readonly Parser<char, Unit> LookaheadKeyWordEnd = Lookahead(Not(IdRest).Or(End()));
+
+        public static readonly Parser<char, string> Bool = True.Or(False).Between(DQuote).Or(True.Or(False)).Labelled("bool");
+
+        public static readonly Parser<char, string> KeyWords =
+            OneOf(Else, Try(Then), Try(If), Try(In)).Before(LookaheadKeyWordEnd).Labelled("keyword");
+
+        public static readonly Parser<char, string> ReservedWords =
+            OneOf(False, Try(True), Else, Try(Then), Try(If), Try(In)).Before(LookaheadKeyWordEnd).Labelled("reserved");
+
+        public static readonly Parser<char, string> Id = Try(Lookahead(Not(ReservedWords).Or(End()))).Then(IdInternal).Labelled("id");
+
+        public static readonly Tuple<Parser<char, Expr>, Stack<bool>, Stack<bool>> Parser = BuildExprParser();
+
+        public static Result<char, Expr> Parse(string input)
+        {
+            try
+            {
+                var ret = Parser.Item1.Parse(input.Trim());
+                if (Parser.Item2.Count == 0 && Parser.Item3.Count == 0) return ret;
+
+                if (Parser.Item2.Count>0)
+                    throw new InvalidOperationException($"Unbalanced C-style Ternary Operators, still have {Parser.Item2.Count} unmatched.");
+                else
+                    throw new InvalidOperationException($"Unbalanced Python-style Ternary Operators, still have {Parser.Item3.Count} unmatched.");
+            }
+            finally
+            {
+                Parser.Item2.Clear();
+                Parser.Item3.Clear();
+            }
+        }
+
+        private static Tuple<Parser<char, Expr>, Stack<bool>, Stack<bool>> BuildExprParser()
+        {
+            Parser<char, Expr> exprParser = null;
+
+            var listParser = Rec(() => exprParser).Separated(Comma.Before(SkipWhitespaces));
+            var funcHead = Id.Or(If).Between(SkipWhitespaces).Before(LParen.Between(SkipWhitespaces));
+            var funcParser = Try(Lookahead(funcHead)).Then(Map((id, prms) => (Expr) new Function(id, prms), funcHead,
+                listParser.Before(RParen))).Labelled("func");
+            var termParser = funcParser
+                .Or(Try(Bool.Select<Expr>(x => new BoolLit(x)))
+                .Or(Num.Select<Expr>(x => new NumLit(x)))
+                .Or(Str.Select<Expr>(x => new StringLit(x)))
+                .Or(Chr.Select<Expr>(x => new CharLit(x)))
+                .Or(Id.Select<Expr>(x => new Identifier(x)))
+                .Or(Rec(() => listParser)
+                    .Between(LBrace.Then(SkipWhitespaces), SkipWhitespaces.Then(RBrace))
```

Comment:
In my experience you get better mileage from an approach where you consume whitespace _after_ every token. You end up needing to use `Try` less - the parser doesn't commit to a branch just because it saw some whitespace.

I usually define a couple of combinators to wrap up the `Then(SkipWhitespaces)` pattern. Something like:

```csharp
static Parser<char, char> Tok(char c)
    => Tok(Char(c));
static Parser<char, string> Tok(string s)
    => Tok(String(s));
static Parser<char, T> Tok<T>(Parser<T> parser)
    => Try(parser).Then(SkipWhitespaces);

static Parser<char, string> Reserved(string s)
    => Reserved(String(s));
static Parser<char, string> Reserved(char c)
    => Reserved(Char(c));
static Parser<char, T> Reserved(Parser<T> parser)
    => Try(parser.Then(Whitespace)).Then(SkipWhitespaces);
```

Then you just wrap all of the terminals in your language in either `Tok` or `Reserved` (depending whether the following whitespace should be compulsory, as in eg the `class` keyword) and forget about whitespace altogether.

```csharp
public static readonly Parser<char, char> LBrace = Tok('{');
public static readonly Parser<char, char> RBrace = Tok('}');
public static readonly Parser<char, char> LParen = Tok('(');
public static readonly Parser<char, char> RParen = Tok(')');
```

(Obviously this doesn't generalise straightforwardly to whitespace-sensitive languages like Python. I've been meaning to add some tools for indentation-based parsing.)